esphome/pins.py

```diff
@@ -161,3 +212,49 @@ def check_strapping_pin(conf, strapping_pin_list, logger):
     # mitigate undisciplined use of strapping:
     if num not in strapping_pin_list and conf.get(CONF_IGNORE_STRAPPING_WARNING):
         raise cv.Invalid(f"GPIO{num} is not a strapping pin")
+
+
+GPIO_STANDARD_MODES = (
+    CONF_ANALOG,
+    CONF_INPUT,
+    CONF_OUTPUT,
+    CONF_OPEN_DRAIN,
+    CONF_PULLUP,
+    CONF_PULLDOWN,
+)
+
+
+def gpio_validate_modes(value):
+    if not value[CONF_INPUT] and not value[CONF_OUTPUT]:
+        raise cv.Invalid("Mode must be input or output")
+    return value
+
+
+def gpio_base_schema(
+    pin_type,
+    number_validator,
+    modes=GPIO_STANDARD_MODES,
+    mode_validator=gpio_validate_modes,
+    invertable=True,
+):
+    """
+    Generate a base gpio pin schema
+    :param pin_type: The type for the pin variable
+    :param number_validator: A validator for the pin number
+    :param modes: The available modes, default is all standard modes
+    :param mode_validator: A validator function for the pin mode
+    :param invertable: If the pin supports hardware inversion
+    :return: A schema for the pin
+    """
+    mode_dict = dict(map(lambda m: (cv.Optional(m, default=False), cv.boolean), modes))
+    schema = cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(pin_type),
+            cv.Required(CONF_NUMBER): number_validator,
+            cv.Optional(CONF_OTHER_USES): cv.boolean,
```

Comment:
Maybe a better name?
```suggestion
            cv.Optional(CONF_ALLOW_OTHER_USES): cv.boolean,
```